### ISSUE TITLE
Auto-disposable tasks for IScheduledExecutor

### ISSUE THREAD

**Issue by tkountis** [hazelcast/hazelcast#16994]
Submitted tasks on `IScheduledExecutor` acquire and hold resources past the lifetime of the job/task. As per the design, each Future returned by any `IScheduledExecutor` API, must be disposed ( `dispose()`) to release the resources on the partition.

The reason this happens is the API itself which returns a `Future` object, and due to the distributed nature of the application, its not easy to track the reference of the `ScheduledFuture` and whether its in use or not. This means that if you indeed hold a reference to that, and we auto-dispose, you
will end up with a "dead" Future object.

Regardless if the task is a single execution or repeatable/periodic one, the rule applies the same.

As with any resource management (File, Connections etc), releasing them, is trivial to forget and not always easy to control; depending on the use-case. We had a few recommendations to support `fire-and-forget` tasks, tasks that once their job is done, they are automatically disposed. Its an idea we have been considering and it was a more natural fit in a world where all the Executors were merged into one, but never got prioritized.

**Comment by tkountis** [comment]
For a work around, we usually suggest an orchestrator design, where a periodic task is introduced that holds a reference to all scheduled tasks, and periodically checks if they are done, in which case it disposes them. An alternative to that, would be for the callers to submit a dispose task to the DurableExecutorService and get the current task disposed.

**Comment by peterjot** [comment]
Hey @Holmistr Could you assign, please?

**Comment by Holmistr** [comment]
@peterjot Sure!
@tkountis Would you mind providing some more details where to start?

**Comment by tkountis** [comment]
Hey @peterjot nice choice, let me know if you need any help throughout your work on this.
I would suggest by starting looking at the lifecycle of a `Runnable`/`Callable` inside the `IScheduledExecutor`, https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/TaskRunner.java#L59 and what would mean to have a task that is self-disposable. 
Once you work this out, it will be easier to review the API side and decide on the changes/additions.
Always keep in mind that its a distributed service, and each task will most likely have 1 or more backups on other nodes / members of the cluster, so this auto-disposal functionality needs to be propagated accordingly.

**Comment by tkountis** [comment]
Hey @peterjot did you manage to play with this? Feel free to reach out if you need a rubber duck.

**Comment by peterjot** [comment]
Hey @tkountis, Thanks for your important tips and sorry for the late answer. I have been thinking on that... My idea is to split `IScheduledFuture` to standard and auto-disposable Future. I think we could add a auto-disposable flag to `TaskDefinition` and instead of returning `IScheduledFuture` in `ScheduledExecturServiceProxy` we could return our brand new Future e.g. `IAutoDisposableScheduledFuture ` when scheduling the auto-disposable task. 

If `TaskDefinition` is auto-disposable maybe we could create something like `AutoDisposableTaskRunner` in that method `com/hazelcast/scheduledexecutor/impl/ScheduledExecutorContainer.java:439` which could be `HazelcastInstanceAware` (to dispose the task after call) instead of `TaskRunner`.

I have been thinking about a new API for auto-disposable scheduling in `IScheduledExecutorService`. Can we do any breaking changes here?

WDYT?

**Comment by tkountis** [comment]
Thanks for the update. Yes your thinking seems more or less correct.
A couple of points to consider

- API wise we can't have breaking changes, but the good part is that we probably don't need them either. I am thinking that if you need an auto-disposable task, we are looking at a schedule-once (& forget) task, not repeatable ones. In which case, we only need one new flavor of the API, which is not much.

- I would probably make use of the TaskDefinition flag (eg. isAutodisposable) to release the resources when we sync-state between members, instead of introducing more operations. Both the primary member and all the replicas know about the definition, so once they receive the sync-state, they can decide whether to drop the task from their internal tracker or not.

Does this make sense?

**Comment by peterjot** [comment]
Hey
I think that we could add methods like `scheduleAndForget`, `scheduleOnMemberAndForget`, `scheduleOnKeyOwnerAndForget`, `scheduleOnAllMembersAndForget` to `IScheduledExecutorService` which create auto-disposable `TaskDefinition`. You're right, removing the tasks during state sync would be defenitly better. Thanks for the correction!

**Comment by mmedenjak** [comment]
I'd prefer to add these as marker annotations to tasks, otherwise we are facing API bloat - a method for every combination of `schedule`, `onKeyMember`, `onOwner`, `onAllMembers` and `andForget`.

**Comment by peterjot** [comment]
@mmedenjak thanks for the tips. I've introduced a method similar to `named()`. It would look like this:
```
Callable task = autoDisposable(named("name", new PlainCallableTask());
service.scheduleOnKeyOwner(task, keyOwner, 0, TimeUnit.SECONDS);
```
